Userland/Libraries/LibCrypto/BigInt/SignedBigInteger.cpp

```diff
@@ -334,6 +335,155 @@ bool SignedBigInteger::operator>=(SignedBigInteger const& other) const
     return !(*this < other);
 }
 
+SignedBigInteger::CompareResult SignedBigInteger::compare_to_double(double value) const
+{
+    VERIFY(!isnan(value));
+
+    if (isinf(value)) {
+        bool is_positive_infinity = __builtin_isinf_sign(value) > 0;
+        return is_positive_infinity ? CompareResult::DoubleGreaterThanBigInt : CompareResult::DoubleLessThanBigInt;
+    }
+
+    bool bigint_is_negative = m_sign;
+
+    bool value_is_negative = value < 0;
+
+    if (value_is_negative != bigint_is_negative)
+        return bigint_is_negative ? CompareResult::DoubleGreaterThanBigInt : CompareResult::DoubleLessThanBigInt;
+
+    // Value is zero, and from above the signs must be the same.
+    if (value == 0.0) {
+        VERIFY(!value_is_negative && !bigint_is_negative);
+        // Either we are also zero or value is certainly less than us.
+        return is_zero() ? CompareResult::DoubleEqualsBigInt : CompareResult::DoubleLessThanBigInt;
+    }
+
+    // If value is not zero but we are, then since the signs are the same value must be greater.
+    if (is_zero())
+        return CompareResult::DoubleGreaterThanBigInt;
+
+    constexpr u64 mantissa_size = 52;
+    constexpr u64 exponent_size = 11;
+    constexpr auto exponent_bias = (1 << (exponent_size - 1)) - 1;
+    union FloatExtractor {
+        struct {
+            unsigned long long mantissa : mantissa_size;
+            unsigned exponent : exponent_size;
+            unsigned sign : 1;
+        };
+        double d;
+    } extractor;
+
+    extractor.d = value;
+    VERIFY(extractor.exponent != (1 << exponent_size) - 1);
+    // Exponent cannot be filled as than we must be NaN or infinity.
+
+    i32 realExponent = extractor.exponent - exponent_bias;
+    if (realExponent < 0) {
+        // |value| is less than 1, and we cannot be zero so if we are negative
+        // value must be greater and vice versa.
```

Comment:
(These are the type of comments I mean)
I don't really feel like writing a proper explanation like for NaN boxing so I can remove them but this is kind of complicated so we should probably keep something of an explanation.